### ISSUE TITLE
Move slang-test and render-test to use the new compilation API

### ISSUE THREAD

**Issue by csyonghe** [shader-slang/slang#4760]
All C functions + the ICompileRequest interface is the old API.
We should move to use COM interfaces directly, and not use the ICompileRequest interface.

**Comment by jkwak-work** [comment]
I am very interested in working on this.
It will be a good chance for me to learn more about our reflection APIs, which I am not very familiar at this point.

**Comment by csyonghe** [comment]
I think this is more about the compilation API, the reflection logic can most likely remain unchanged. But this is a good opportunity to get familiar with the compilation API.

**Comment by aleino-nv** [comment]
> I am very interested in working on this. It will be a good chance for me to learn more about our reflection APIs, which I am not very familiar at this point.

@jkwak-work Would it be ok if I took this task, or are you planning to get to it soon?

**Comment by aleino-nv** [comment]
@csyonghe What's meant by "new compilation API"? COM-based rather than C API?

**Comment by aleino-nv** [comment]
@csyonghe Does the non-deprecated API provide anything like `ICompileRequest::compile`?

There is one usage of `ICompileRequest` in `slang-test`, and that's this line in `slangc-tool.cpp`:

https://github.com/shader-slang/slang/blob/0fee8c16e25521d13b50427ac6bf87d7f3419b9c/tools/slang-test/slangc-tool.cpp#L58

**Comment by csyonghe** [comment]
That use is fine, because `slangc` is currently still on legacy API, and we just have to keep it that way.

The main body of work is to get render-test to use the new API, which amounts to changing `ShaderCompilerUtil::_compileProgramImpl` function to not use any `sp*` APIs or the `ICompileRequest` API.

**Comment by aleino-nv** [comment]
Ok, I worked on render-test today. I've almost got a sketch of a patch ready.

**Comment by aleino-nv** [comment]
I had to debug quite a bit today, because some reflection data seems to have broken with my patch.
I created a simpler clean-up patch which is now in review, as a stepping stone.

I still need at least a day on this task to actually get rid of the compile request stuff from render-test.

**Comment by aleino-nv** [comment]
I created https://github.com/shader-slang/slang/pull/6440 today.

However it runs into a seemingly pre-existing issue where there's a crash in loadModuleFromSourceString if you call it again from a second session. I will file a bug for this tomorrow and try to fix that issue first.

**Comment by aleino-nv** [comment]
> However it runs into a seemingly pre-existing issue where there's a crash in loadModuleFromSourceString if you call it again from a second session. I will file a bug for this tomorrow and try to fix that issue first.

This issue is somehow gone today after syncing and rebuilding. I wonder if I ran into some build issue...

I'll fix some build issues and see if the patch passes full testing now.

**Comment by aleino-nv** [comment]
There are still some tests that fail. tests/bugs/texture2d-gather.hlsl fails because findEntryPointByName fails to find the entry point.
When I mark out the entry point explicitly with e.g. [shader("vertex")], then the test passes on all API's except DX11 and DX12.

According to this: 
> Slang supports code that leaves off [shader(...)] attributes; in these cases application developers must specify the names and stages for their entry points via explicit command-line or API options.
https://shader-slang.org/slang/user-guide/conventional-features.html#functions

It should also be possible to use API options to mark out entry points, so I tried adding the entry points via slang::CompilerOptionName::EntryPointName but still findEntryPointByName fails.

I'm now investigating if there are pre-existing bugs around entry point lookup...

**Comment by aleino-nv** [comment]
There seems to be a pre-existing API bug where findEntryPointByName fails when using the new session API rather than compile requests.

Since some tests don't use the `shader["..."]` syntax to explicitly point out entry points, I believe the only option is to first fix this issue.

I filed https://github.com/shader-slang/slang/issues/6452 which I am now working on before my patch can pass.

**Comment by aleino-nv** [comment]
Thanks Yong, I changed https://github.com/shader-slang/slang/issues/6452 to be about just clarifying this in the user guide and function documentation comment for findEntryPointByName.

**Comment by aleino-nv** [comment]
There are still some tests failing. I debugged it and found an issue with argument parsing in the new API: https://github.com/shader-slang/slang/pull/6461

**Comment by aleino-nv** [comment]
Found another issue where new session API doesn't respect the matrix options https://github.com/shader-slang/slang/pull/6476, leading tests to fail when migrating.

**Comment by aleino-nv** [comment]
@csyonghe How should we handle when -load-repro is passed as an argument in SlangResult _compileProgramImpl?

Currently the repro loading logic is only implemented for the deprecated compile request objects.
If I use IGlobalSession::parseCommandLineArguments then the repro gets loaded into a temporary compile request object `tempReq`, and then that is discarded as soon as the session description has been created, along with the entry point information that I suppose came from the repro.

I could easily change things so that when the load repro argument is parsed it's actually added to the session description options, so that we can thread the repro name through to ISession creation time. (During session creation we can then call the same compile request -based function that already exists.)

Is that the desired fix?

**Comment by csyonghe** [comment]
We can fallback to use the existing ICompileRequest path if we see a -load-repro. It is not implementable with the new API.

**Comment by aleino-nv** [comment]
The next issue is that [some tests are using legacy commands for specifying global/entrypoint existential types](https://discord.com/channels/1303735196696445038/1303735244108595330/1344954145156890656).

Some tests using these commands are passing with the new compilation API, but others are failing.

I noticed that the new IComponentType::getSpecializationParamCount API returns 0 for https://github.com/shader-slang/slang/blob/a09d554721ecf0c6d967d5f55b3416e3284e71f2/tests/compute/dynamic-dispatch-bindless-texture.slang, whereas I would have expected 1 after debugging and inspecting the returned value from other similar tests.

**Comment by aleino-nv** [comment]
@csyonghe I also noticed that tests that use `__import` (e.g. tests/render/cross-compile0.hlsl) are failing to resolve the symbols they import, with the new API. The import itself seems to work fine after I added the full path to the test when loading the module.

Is `__import` documented anywhere? Is it supposed to work with the new compilation API?
I haven't investigated too closely yet, just thought I'd check before spending time on it, in case it's also considered deprecated.

**Comment by aleino-nv** [comment]
Update: I currently have 93% (4311/4607) of tests passing with the new API.
Lots of the failures are for CUDA targets, so I will investigate those next.

**Comment by aleino-nv** [comment]
While debugging the CUDA failures, I noticed that for some reason the [new compilation API generates invalid code when targeting PTX](https://github.com/shader-slang/slang/issues/6507).

Switching to fix that issue before continuing with this one.

**Comment by aleino-nv** [comment]
> I currently have 93% (4311/4607) of tests passing with the new API. Lots of the failures are for CUDA targets, so I will investigate those next.

After fixing the CUDA codegen issue, now 98% (4527/4618) of the tests are passing.
Looking into those next...

**Comment by aleino-nv** [comment]
Option set serialization code was broken, causing downstream compilers to be invoked with invalid arguments.

Small fix here: https://github.com/shader-slang/slang/pull/6515

This should take care of lots of failures.

**Comment by aleino-nv** [comment]
Found and fixed lots more instances of [a previous issue](https://github.com/shader-slang/slang/issues/6507) where `lowerVaryingInputAsConstRef` was not being set because of missing [shader(...)] attributes in the tests.

In this case I added [shader(...)] attributes to the tests.

**Comment by aleino-nv** [comment]
Update: 99% (4582/4618) of tests are passing with the new API.

Continuing to look into failures...

**Comment by aleino-nv** [comment]
I have a fix for another option parsing bug: https://github.com/shader-slang/slang/pull/6530 that caused 10 of the test failures.
Pass rate is now (4592/4619).

**Comment by aleino-nv** [comment]
I fixed some tests by making sure that warnings were printed to stderr (some tests verify that warnings are emitted in some cases).
However that broke loads of other tests because the stderr sometimes contains "Unhandled Exception", so now I need to debug that.

**Comment by aleino-nv** [comment]
The unhandled exception stuff was just due to a missing nullptr check.

Then I found that lots of tests started failing because they contained warnings like 
```
warning 41012: entry point 'computeMain' uses additional capabilities that are not part of the specified profile...
```

It turns out that the new session-based API triggers this warning by adding "unknown" profile option to the option set, whereas the old compile request -based API does not, and therefore doesn't trigger the warning.

Proposed fix here: https://github.com/shader-slang/slang/pull/6538.

**Comment by aleino-nv** [comment]
The next issue is [this bool in `FrontEndCompileRequest::checkEntryPoints`](https://github.com/shader-slang/slang/blob/e1952dc8cf8f5b62d00ce114e353c5390cc6c37a/source/slang/slang-check-shader.cpp#L945):

```
    // The validation of entry points here will be modal, and controlled
    // by whether the user specified any entry points directly via
    // API or command-line options.
    //
    // TODO: We may want to make this choice explicit rather than implicit.
    //
    // First, check if the user requested any entry points explicitly via
    // the API or command line.
    //
    bool anyExplicitEntryPoints = getEntryPointReqCount() != 0;
```

With the old compilation API, this bool is `true` and that leads to detecting an appropriate profile for various entrypoints.
With the new compilation API, the bool is `false` and that leads to selecting a profile that doesn't have the required capabilities in lots of tests.

The issues in the new API happen in `loadModuleFromSourceString`, where we don't yet know about entrypoints.

The test I'm currently focusing on `tests/pipeline/rasterization/mesh/task-groupshared.slang` and I can make that pass by adding `[shader("amplification")]`, so I'll just do that for now.

**Comment by aleino-nv** [comment]
The next issue is 

"unresolved extern" errors for tests that add modules via command line options, namely these:

1. tests/serialization/obfuscated-serialized-module-test.slang.1 (dx11)
2. tests/serialization/obfuscated-serialized-module-test.slang.2 syn (wgpu)
3. tests/serialization/obfuscated-serialized-module-test.slang.3 syn (dx12)
4. tests/serialization/serialized-module-test.slang.1 (dx11)
5. tests/serialization/extern/extern-test.slang.2 (dx11)
6. tests/serialization/serialized-module-test.slang.2 syn (wgpu)
7. tests/serialization/extern/extern-test.slang.3 syn (wgpu)
8. tests/serialization/serialized-module-test.slang.3 syn (dx12)
9. tests/serialization/extern/extern-test.slang.4 syn (dx12)

What goes wrong is again related to the fact that `Session::parseCommandLineArguments` just parses arguments into the old compile request, which will just add the modules to the request without adding anything to the actual options being produced.
In other words, the module dependency just gets dropped on the floor and so we end up with unresolved externals.

I believe in this case what we want to do is this:
1. Modify `EndToEndCompileRequest::processCommandLineArguments` such that it will also add the options, in addition to the existing side-effect. I don't think adding the options will hurt anything down the line, and the original side effect should probably be removed when the old API is completely removed.
2. In render-test when creating the session, we look for these options and create IModules from them, which we then add to the composite component.

@csyonghe Does that sound right?

**Comment by aleino-nv** [comment]
Here is my PR for part 1: https://github.com/shader-slang/slang/pull/6539

I haven't done part 2 yet, but that's next.

**Comment by aleino-nv** [comment]
Update: Pass rate is 4599/4619 with 19 unexpected failures, listed below.

These fall into three categories:

1. Failures related to 'Dynamic specification of generic parameters'
 - I'm expecting to fix this by fixing the tests themselves to use new features related to this
2. Failures outputting "error: External function used in non-library profile"
 - I haven't investigated yet.
3. "unresolved extern" errors
 - This is what I've just debugged. I think I'll have a fix soon.

```
// Dynamic specification of generic parameters
tests/compute/dynamic-dispatch-bindless-texture.slang.1 (cuda)
tests/compute/interface-shader-param.slang (dx11)
tests/compute/interface-shader-param.slang.1 (dx12)
tests/compute/interface-shader-param.slang.2 (vk)
tests/compute/interface-shader-param.slang.6 syn (cuda)
tests/language-feature/shader-params/interface-shader-param-ordinary.slang (dx11)
tests/language-feature/shader-params/interface-shader-param-ordinary.slang.1 (dx12)
tests/language-feature/shader-params/interface-shader-param-ordinary.slang.2 (vk)

// error: External function used in non-library profile
tests/library/library-test.slang.2 (dx12)
tests/library/export-test.slang.2 (dx12)

// "unresolved extern" errors (currently debugging)
tests/serialization/obfuscated-serialized-module-test.slang.1 (dx11)
tests/serialization/obfuscated-serialized-module-test.slang.2 syn (wgpu)
tests/serialization/obfuscated-serialized-module-test.slang.3 syn (dx12)
tests/serialization/serialized-module-test.slang.1 (dx11)
tests/serialization/extern/extern-test.slang.2 (dx11)
tests/serialization/serialized-module-test.slang.2 syn (wgpu)
tests/serialization/extern/extern-test.slang.3 syn (wgpu)
tests/serialization/serialized-module-test.slang.3 syn (dx12)
tests/serialization/extern/extern-test.slang.4 syn (dx12)
```

**Comment by aleino-nv** [comment]
These ones are now resolved as described above:

```
// "unresolved extern" errors (currently debugging)
tests/serialization/obfuscated-serialized-module-test.slang.1 (dx11)
tests/serialization/obfuscated-serialized-module-test.slang.2 syn (wgpu)
tests/serialization/obfuscated-serialized-module-test.slang.3 syn (dx12)
tests/serialization/serialized-module-test.slang.1 (dx11)
tests/serialization/extern/extern-test.slang.2 (dx11)
tests/serialization/serialized-module-test.slang.2 syn (wgpu)
tests/serialization/extern/extern-test.slang.3 syn (wgpu)
tests/serialization/serialized-module-test.slang.3 syn (dx12)
tests/serialization/extern/extern-test.slang.4 syn (dx12)
```

**Comment by aleino-nv** [comment]
Pass rate is now 4608/4619.

There are two remaining groups of failures:

```
// Dynamic specification of generic parameters
tests/compute/dynamic-dispatch-bindless-texture.slang.1 (cuda)
tests/compute/interface-shader-param.slang (dx11)
tests/compute/interface-shader-param.slang.1 (dx12)
tests/compute/interface-shader-param.slang.2 (vk)
tests/compute/interface-shader-param.slang.6 syn (cuda)
tests/language-feature/shader-params/interface-shader-param-ordinary.slang (dx11)
tests/language-feature/shader-params/interface-shader-param-ordinary.slang.1 (dx12)
tests/language-feature/shader-params/interface-shader-param-ordinary.slang.2 (vk)

// error: External function used in non-library profile
tests/library/library-test.slang.2 (dx12)
tests/library/export-test.slang.2 (dx12)
```


I believe resolving the former group requires updating the tests to using newer test framework features that are not documented, so I need to get some information about those first.

I'm looking at the latter group first. (The `error: External function used in non-library profile` failures.)

These tests are adding DXIL modules for use during downstream linking.
I see that these DXIL modules get added, during option parsing, as artifacts to the linkage, which is part of the obsolete compilation request interface. 
In the non-deprecated API, I see that we have `slang::CompilerOptionName::DownstreamArgs`. 
I will look into whether such options can be used to achieve the same thing.

**Comment by aleino-nv** [comment]
I found that the new compilation API had no way of linking against DXIL code, so I created [this PR](https://github.com/shader-slang/slang/pull/6569) that resolves the "error: External function used in non-library profile".
The pass rate with this PR is now 4610/4619.

Now there is just one last group of failing tests:

```
// All these are related to test infrastructure for dynamic specification of generic parameters
tests/compute/dynamic-dispatch-bindless-texture.slang.1 (cuda)
tests/compute/interface-shader-param.slang (dx11)
tests/compute/interface-shader-param.slang.1 (dx12)
tests/compute/interface-shader-param.slang.2 (vk)
tests/compute/interface-shader-param.slang.6 syn (cuda)
tests/language-feature/shader-params/interface-shader-param-ordinary.slang (dx11)
tests/language-feature/shader-params/interface-shader-param-ordinary.slang.1 (dx12)
tests/language-feature/shader-params/interface-shader-param-ordinary.slang.2 (vk)
```

I'll try to figure out how these are supposed to be updated. As said before, there may already be test infrastructure for this, but it's not documented yet.

**Comment by aleino-nv** [comment]
Closed by https://github.com/shader-slang/slang/pull/6514.
